### ISSUE TITLE
Ruby 2.5.0, fastly-rails 0.8.0 errors

### ISSUE THREAD

**Issue by tyrauber** [fastly/fastly-rails#80]
In upgrading our Rails App, I also updated the fastly-rails gem from 0.2 to 0.8.0. But in doing so, I am  now seeing a whole string of new failures related to the fastly-rails gem.  Roughly ~800 failures in a 24 hour period from a single worker. We also updated Ruby from 2.3.1 to 2.5.0.  Perhaps, related?

Environment:

- Rails 5.1.5
- Ruby 2.5.0
- fastly-rails 0.8.0

Here is the assortment of errors we are seeing with relevant stacktraces.

Anyone else seeing anything like this?  Any suggestions?

- **JSON::ParserError: Empty input at line 1, column 1 [parse.c:926] in 'HTTP/1.1 200 OK Content-Type: application/json C**
```
1 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 124 in parse
2 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 124 in post_and_put
3 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 87 in post
4 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/service.rb" line 74 in purge_by_key
5 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/client.rb" line 12 in purge_by_key
6 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails.rb" line 26 in purge_by_key
7 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/active_record/surrogate_key.rb" line 37 in purge
```

- **Net::HTTPBadResponse: wrong status line: "ntent-Length: 50”**
```
1 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 42 in read_status_line
2 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 29 in read_new
3 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1494 in block in transport_request
4 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1491 in catch
5 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1491 in transport_request
6 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1464 in request
16 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1478 in send_entity
17 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1266 in post
18 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 122 in post_and_put
19 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 87 in post
20 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/service.rb" line 74 in purge_by_key
21 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/client.rb" line 12 in purge_by_key
22 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails.rb" line 26 in purge_by_key
23 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/active_record/surrogate_key.rb" line 37 in purge
```
- **Errno::EBADF: Bad file descriptor**
```
1 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 181 in wait_readable
2 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 181 in rbuf_fill
3 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 125 in read
4 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 293 in block in read_body_0
5 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 278 in inflater
6 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 283 in read_body_0
7 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 204 in read_body
8 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1479 in block in send_entity
9 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1503 in block in transport_request
10 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 165 in reading_body
11 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1502 in transport_request
12 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1464 in request
20 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1478 in send_entity
21 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1266 in post
22 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 122 in post_and_put
23 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 87 in post
24 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/service.rb" line 74 in purge_by_key
25 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/client.rb" line 12 in purge_by_key
26 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails.rb" line 26 in purge_by_key
27 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/active_record/surrogate_key.rb" line 37 in purge
```
- I**OError: closed stream**
```
1 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 181 in wait_readable
2 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 181 in rbuf_fill
3 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 157 in readuntil
4 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 167 in readline
5 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 40 in read_status_line
6 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 29 in read_new
7 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1494 in block in transport_request
8 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1491 in catch
9 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1491 in transport_request
10 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1464 in request
20 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1478 in send_entity
21 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1266 in post
22 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 122 in post_and_put
23 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 87 in post
24 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/service.rb" line 74 in purge_by_key
25 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/client.rb" line 12 in purge_by_key
26 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails.rb" line 26 in purge_by_key
27 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/active_record/surrogate_key.rb" line 37 in purge
```

**Comment by mrthan** [comment]
I'm running exactly these versions with no issues, I suspect it might be something in that aws-sdk-core gem patching something in Net::Http causing issues (https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/seahorse/client/net_http/patches.rb).

Are you able to alter the version of AWS SDK or even remove it to test the Fastly interaction?

**Comment by tyrauber** [comment]
Ugh, aws-sdk-core monkeypatching Net::HTTP. Nice.  We need the aws-sdk for managing assets on S3, but I am going to set the gem to require false and see if that limits the scope and fallout.  I'll report back. Thanks for the direction.

**Comment by tyrauber** [comment]
Unfortunately, setting require: false doesn't isolate the patch. Calling require 'aws-sdk-s3' in a class or module still invokes and applies the patch globally.  I have posted to aws-sdk-ruby:
https://github.com/aws/aws-sdk-ruby/issues/1752

**Comment by tyrauber** [comment]
@mrthan Unfortunately, Amazon monkeypatching Net::HTTP was a red herring.  They removed the Net::HTTP patch for ruby 2.5.0 and pushed out an updated aws-ruby gem.  I updated the gem in our environment, but I am still seeing these errors (just no longer with aws-sdk-ruby in the stacktrace).  I updated the stacktraces above.

Any other ideas?  We are doing purges in a Sidekiq. They work 90% of the time.  I know we issue a lot of purge requests? Could it be a result rate limiting? It seems like they fail in clusters. I'll see 9-10 failures happen at a time.

**Comment by tyrauber** [comment]
It actually looks like the root issue is in fastly-1.15.0. Should I post this issue there instead?

**Comment by mrthan** [comment]
I'd say yes, post it over there. But do you have updated stack traces now that the AWS change is in place?

If it was rate limiting in effect I'd hope it was a nicer error/rejection rather that an invalid response.

Aside: I tend to use mostly tag based invalidation, is that an option to group content to allow for fewer individual purge requests?

**Comment by tyrauber** [comment]
@mrthan Yeah, sorry, I edited my first comment and updated the traces. Note: aws-sdk-ruby is no longer present in the stacktrace.  I posted to [fastly-ruby](https://github.com/fastly/fastly-ruby/issues/130) also.

We use table_key or record_key for our surrogate keys depending on action type. Fortunately/Unfortunately, our app has a lot of user interaction - saving, bookmarking, following - which requires multiple records be purged to keep the content up-to-date.  We try to be conservative about the amount that we purge and what we purge. Obviously our preference is to be serving from cache.  But, I'll re-evaluate our purge architecture to see if there is anything we can do to minimize the frequency we purge.
